pennylane/transforms/decompositions/single_qubit_unitary.py

```diff
@@ -81,27 +105,19 @@ def zyz_decomposition(U, wire):
     """
     U = _convert_to_su2(U)
 
-    # Check if the matrix is diagonal; only need to check one corner.
-    # If it is diagonal, we don't need a full Rot, just return an RZ.
-    if math.allclose(U[0, 1], [0.0]):
-        omega = 2 * math.angle(U[1, 1])
-        return [qml.RZ(omega, wires=wire)]
-
-    # If the top left element is 0, can only use the off-diagonal elements. We
-    # have to be very careful with the math here to ensure things that get
-    # multiplied together are of the correct type in the different interfaces.
-    if math.allclose(U[0, 0], [0.0]):
-        phi = 0.0
-        theta = -np.pi
-        omega = 1j * math.log(U[0, 1] / U[1, 0]) - np.pi
-    else:
-        # If not diagonal, compute the angle of the RY
-        cos2_theta_over_2 = math.abs(U[0, 0] * U[1, 1])
-        theta = 2 * math.arccos(math.sqrt(cos2_theta_over_2))
-
-        el_division = U[0, 0] / U[1, 0]
-        tan_part = math.cast_like(math.tan(theta / 2), el_division)
-        omega = 1j * math.log(tan_part * el_division)
-        phi = -omega - math.cast_like(2 * math.angle(U[0, 0]), omega)
-
-    return [qml.Rot(math.real(phi), math.real(theta), math.real(omega), wires=wire)]
+    if math.is_abstract(U):
+        interface = math._multi_dispatch(U)
+
+        if interface == "jax":
```

Comment:
@josh146 This is one version of the `jax.lax.cond` implementation I have tried. It deliberately leaves out the `wires`, but throws an error complaining that functions are not valid JAX types.

I have also tried this:
```python
        if interface == "jax":
            return cond(
                math.allclose(U[0, 1], 0.0),
                lambda x: [qml.RZ(2 * math.angle(x[1, 1]), wires=wire)],
                lambda x: [qml.Rot(*_compute_rot_angles(x), wires=wire)],
                U
            )
```
where it complains that `Value RZ(Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=1/1)>, wires=[0]) with type <class 'pennylane.ops.qubit.parametric_ops.RZ'> is not a valid JAX type`. (Can see here though that the wires argument is set correctly and not with a traced value).

Finally, I thought to instead have the conditional return only the angles, but `cond` cannot seem to handle the case where the two branches return different types, which is the case here (array with 1 element for an `RZ` vs. 3 elements for `Rot`).

Reply:
Ah, I think I see what is happening --- both of the branches for `cond` must be functions that _accept and return_ JAX types. So the issue here is that `qml.RZ` and `qml.Rot` are not JAX types!

I managed to get the following working:
```python
if interface == "jax":

    angles = cond(
        math.abs(U[0, 1]) < 1e-5,
        lambda x: (2 * math.angle(x[1, 1]), 0., 0.),
        lambda x: _compute_rot_angles(x),
        U
    )

    return [qml.Rot(*angles, wires=wire)]
```

This works because both functions return `tuple[jax.DeviceArray[float32]]`. Note that both had to return a tuple of floats; if the shape of the function returns differed, JAX complained.

The bit I can't work out then --- how to tell whether to apply `Rot` or `RZ` 😬 Since both return a 3-tuple of floats!

I tried doing this:

```python
if interface == "jax":

    RZ_gate, angles = cond(
        math.abs(U[0, 1]) < 1e-5,
        lambda x: [True, (2 * math.angle(x[1, 1]), 0., 0.)],
        lambda x: [False, _compute_rot_angles(x)],
        U
    )

    if RZ_gate:
        return qml.RZ(angles[0], wires=wire)
    return [qml.Rot(*angles, wires=wire)]
```
but this also fails (this time at the `if` statement), because `RZ_gate` becomes an abstract `bool` array 😢

Reply:
> Ah, I think I see what is happening --- both of the branches for cond must be functions that accept and return JAX types. 
> So the issue here is that qml.RZ and qml.Rot are not JAX types!

Despite the error message saying exactly that, it didn't make sense until I read this sentence :woman_facepalming: If even you could not figure out a way around it, we might just have to leave this as-is. It is still completely jittable, at least. Though I think this precludes trying anything similar for the `two_qubit_decomposition`, there are way more conditional statements in there (I was at least able to speed it up a bit, with all the other improvements).

Reply:
I really wish I could think of something 😢 We need to ask a JAX expert, because I feel like this _should_ be possible.

Reply:
Do you know one, and/or have connections with any of the devs?

Reply:
I don't unfortunately, but we could try asking over at https://github.com/google/jax/discussions?

Reply:
@josh146 whoa, this link is gold!! :100: :star:  There are actually a lot of similar questions, so let me see first if we can find our solution there already.

Reply:
Okay based on some discussions, I've cobbled together the following which still fails, so I'm increasingly convinced this isn't possible.

```python
from jax.tree_util import Partial

def _make_rz(U):
    # Using Partial ensures return is a valid JAX type
    return Partial(qml.RZ, *[2 * math.angle(U[1, 1]), 0.0, 0.0]) 

def _make_rot(U):
    return Partial(qml.Rot, *_compute_rot_angles(U))

f = cond(
    math.allclose(U[0, 1], 0.0),
    _make_rz,
    _make_rot,
    U
)

return f(wires=wire) #Ensures wires is static
```

I modified `RZ` internally to be able to handle multiple parameters being passed, and then take only the first one, so that we can return lists of floats of the same length (but this makes me very uncomfortable!). However, we still get

```
TypeError: true_fun and false_fun output must have same type structure, got PyTreeDef(CustomNode(<class 'jax._src.tree_util.Partial'>[<class 'pennylane.ops.qubit.parametric_ops.RZ'>], [(*, *, *), {}])) and PyTreeDef(CustomNode(<class 'jax._src.tree_util.Partial'>[<class 'pennylane.ops.qubit.parametric_ops.Rot'>], [(*, *, *), {}])).
```
because the two gates are different types.

Similarly I have tried:

```python
def _make_rz(U):
    return [2 * math.angle(U[1, 1]), 0.0, 0.0]

def _make_rot(U):
    return [*_compute_rot_angles(U)]

index = cond(
    math.allclose(U[0, 1], 0.0),
    lambda _: 0,
    lambda _: 1, 
    operand=None
)

return switch(
    index,
    [
        lambda x : Partial(qml.RZ, _make_rz(x)),
        lambda x : Partial(qml.Rot, _make_rot(x))
    ],
    U
)(wires=wire)
```
which has the same problem, but with the  `switch` statement instead of `cond`. I've also tried adding something like this:

```python
gate_map = {
    0 : qml.RZ,
    1 : qml.Rot
}
```
to try and catch the output of the `cond` function, but then it fails because the returned value cannot be used as an input to the dictionary.

So yeah... :cry:

Reply:
😢 

The JIT compiler is too smart!

(btw I didn't think `math.allclose(U[0, 1], 0.0)` was JIT compatible, hence why I changed it to `math.abs(U[0, 1]) < 1e-5` above? but it sounds like your error messages were occurring before this line was hit?)

Reply:
>  but it sounds like your error messages were occurring before this line was hit?

Yeah changing to `math.abs` doesn't affect it. I wonder, since `RZ` and `Rot` have the same parent class, is there a way to trick it into thinking they are the same type?